### PR TITLE
Fix inaccurate integrated gradients computation in tutorial

### PR DESCRIPTION
**Pull request by bentyeh**

### CODE REVIEW

**Comment by dougkelly** [comment]
Hi Benjamin,

Thanks for taking the time to submit fixes. Can you please fix the diff so I can review?

**Comment by bentyeh** [comment]
@dougkelly Diff is fixed! Also realized there were some more inaccuracies in the computation which I addressed.

If someone who is very familiar with the integrated gradients method could review the math, that would be great.

The commit comment summarizes the main changes I made:
1. `plot_riemann_sums()` examples now evaluates function at correctly evenly-spaced points.
2. `generate_alphas()` compiles correctly for all methods (types issues) and correctly generates evenly-spaced points for Riemann sums. The compilation is from the `@tf.function` decorator applied to the `integrated_gradients()` function, which calls `generate_alphas()`.
3. `integrated_gradients()` is now robust to batch sizes. Previously, if `m_steps % alpha_batch == 1`, then the function would fail if `method == 'riemann_trapezoidal'` because `integral_approximation()` called by `integrated_gradients()` with the 'riemann_trapezoidal' method requires at least 2 gradient values
4. `integrated_gradients()` now correctly accounts for the area between the last position in a previous batch and the first position in a new batch when using the trapezoidal method
5. `integrated_gradients()` now correctly aggregates integrated gradients across batches.
6. `convergence_check()` now correctly compares sum of attributions to *softmax* scores (thus matching how gradients are computed in `compute_gradients()`).
7. Accurate computation of integrated gradients allows large reduction in number of steps required for examples.
8. Fixed some typos.
9. Added note to the beginning about a shortened version of this notebook available as a [Tensorflow tutorial](https://www.tensorflow.org/tutorials/interpretability/integrated_gradients).
   - If the bugs in this notebook are acknowledged and this pull request is approved, the fixes here should be similarly applied (where applicable) to the Tensorflow tutorial version.

**Comment by dougkelly** [comment]
Hi @bentyeh,

Thanks again for the excellent feedback and taking the time to propose fixes. I would like to move forward with accepting many of your proposed fixes as well as add in a few alternative fixes to address key issues you identified with the goal of 1.) correcting the IG computation across batches 2.) keep the code as simple as possible and in line with the original implementation for teachability. 

Can you please add me to https://github.com/bentyeh/training-data-analyst.git so I can push these changes to the master branch there and they will be added to this PR? I plan to do the same for the associated TF docs PR as well to make sure your name is on the PRs rather than forking them :)

**Comment by bentyeh** [comment]
@dougkelly Thanks for taking a look! I've added you to my forked repos.

**Comment by dougkelly** [comment]
@bentyeh Thanks! I pushed a commit to this PR (see the ReviewNB link above). Have a look and let me know if you have any questions before I merge it. I'll do the same for the TF docs PR next as well.

1. Accepted. 
2. Accepted. Great catch on the integer division instead of the intended float division as well as fix on Riemann midpoint and trapezoidal m_steps+1. For 2-5, I tested with TF 2.1, 2.2, and latest 2.3. Saw very different behavior than expected with the original code with @tf.function and across numpy==1.8.3 and 1.8.5 and your fix worked consistently across these dependencies.
3 & 4 & 5. Thanks for highlighting and proposing fixes; unfortunately both the original and fixes resulted in incorrect IG calculation under certain scenarios with averaging reduction on each batch + averaging makes the added divide by m_steps line after scaling unnecessary. I changed the implementation to batch just the gradient computation, concat the batches into (n_interpolated_images, H, W, 3) gradients tensor, and kept just the averaging reduction for integral approximation. I found this to be more concise than summing the gradients and then dividing by m_steps and also to more closely match the original IG implementation which is helpful from an instructional standpoint. I removed @tf.function as well due to unexpected and inconsistent behavior under different dependencies mentioned above and hard to debug re-tracing e.g. @tf.function resulted in very different numerical results under odd batch sizes under different TF versions; it either keep loop as Python resulting in a lot of retracing for gradient function or converted to tf.TensorArray which didn't go for different batch sizes. I also just added a note to keep the batch size less than m_steps in the docstring to address this scenario. Finally, added .prefetch(tf.data.experimental.AUTOTUNE) for a bit of a performance boost on batching gradient computation.
6. Accepted.
7. Verified. Computation fixes on `generate_alphas()` and batching reduced m_steps needed to converge. 
8. Accepted. Good eye on typo on image height dimension. Added additional typo fixes e.g. Tensorflow -> TensorFlow. Updated doc strings and text accompanying changes where needed.
9. Accepted. Thanks for adding link back.

**Comment by bentyeh** [comment]
@dougkelly Thanks for your thorough testing, since I did not test with different TF versions.

I have 2 concerns regarding point 2.
- Does your observation that using `@tf.function` produces different IG calculations with different dependencies warrant further investigation? Does that imply a TensorFlow bug? Sorry if the question is naive, but I am not super familiar with the internals of TensorFlow, beyond that using `@tf.function` compiles the computation graph for the function it decorates rather than building the graph each time the function is called.
- I don't quite understand what you mean that the fixes I proposed would result
  > in incorrect IG calculation under certain scenarios with averaging reduction on each batch + averaging makes the added divide by m_steps line after scaling unnecessary.

  In my implementation of `integrated_gradients()` (see line 90 in the ReviewNB diff for function `def integrated_gradients()`), I use a `'sum'` reduction and then divide by `m_steps` at the very end, which should give the same result as your approach. I'm not saying my implementation is better (except maybe marginally less memory usage), just that I'm confused why it would give an "incorrect IG calculation."

**Comment by dougkelly** [comment]
Hi @bentyeh, 

1. Yes. There was 2 TensorFlow bugs at play here introduced in TF2.3 reported to the TF autograph team that are now actively being worked on and slated for fixes in TF 2.4. The first involved differences in how autograph (@tf.function) was handling tf.data iteration and batching and its interaction with tf.logging e.g. for batch in (tf.data object) results in different graph code than for batch in iter(tf.data object) that primarily just add latency and occasional warnings on excessive tracing / or deprecation notices but also resulted in just the first batch getting processed under some of the implementations tested. The second involved how autograph warnings and errors interact with tf.logging and %%timeit and %%memit performance profiling in Colab that causes the kernel to hang indefinitely / restart.

2. Apologies I misread your proposed fix initially. Your implementation is absolutely correct and was actually slightly faster and memory equivalent to the Python one I pushed last week in testing. I see you hardcoded the sum reduction in the IG function batch loop and incorrectly thought you were still exposing mean integral approximation on each batch as a function argument which a user could change and would result in incorrect IG computation like the first tutorial version as well as making the / m_steps redundant. 

After profiling out a few alternatives yesterday and consulting with TF autograph folks, we are going to move forward with the solution just pushed for now that addresses IG computation errors you highlighted, is significantly faster with @tf.function at a slight memory cost for TF2.3, avoids potential issues introduced by bugs in #1, while seeking to balance performance and teachability considerations. Thanks for your contributions!
